### PR TITLE
fix: support mapped tasks

### PR DESCRIPTION
**Pull request by james-union**
This issue is to add support for mapped tasks to the Launch Form UX. While mapped tasks are most often one-dimensional (eg, k:v) flyte supports nested mapping (n-dimensional) as well. To support both use-cases we will provide UX that will renders a standard input field for 1D mappings and a text area (eg, where users can paste JSON text) for n-D mappings.

https://user-images.githubusercontent.com/105876962/175947368-c69d07a8-c4b4-410f-b0e8-6291e32afcd0.mov

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Map type was unsupported and there were no inputHelpers for MAP. I added those helpers and enabled the Map type to be supported. The UI is like the above video

## Tracking Issue

fixes https://github.com/flyteorg/flyteconsole/issues/483

### CODE REVIEW

File: packages/zapp/console/src/components/Launch/LaunchForm/MapInput.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  );
+  let valueControl: JSX.Element;
+
+  switch (subtype?.type) {
+    case InputType.String:
+    case InputType.Integer:
+      valueControl = (
+        <TextField
+          error={valueError}
+          id={`${getLaunchInputId(name)}-value`}
```
Comment (anrusina) [inline packages/zapp/console/src/components/Launch/LaunchForm/MapInput.tsx:76]: I see you are adding ids in ost of your components. Just curious what is the reasoning? Is it for unit test find by id?

File: packages/zapp/console/src/components/Launch/LaunchForm/inputHelpers/map.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import { stringifyValue } from 'common/utils';
+import { Core } from 'flyteidl';
+import { InputTypeDefinition, InputValue } from '../types';
+import { getHelperForInput } from './getHelperForInput';
+import { ConverterInput, InputHelper, InputValidatorParams } from './types';
+
+const missingSubTypeError = 'Unexpected missing subtype for map';
```
Comment (anrusina) [inline packages/zapp/console/src/components/Launch/LaunchForm/inputHelpers/map.ts:7]: All strings need to be added to strings file. example:
`packages/zapp/console/src/components/common/strings.ts`
usage:
{t('annotationsHeader')}

File: packages/zapp/console/src/components/Launch/LaunchForm/utils.ts [hunk 3]
```diff
@@ -190,3 +191,16 @@ export function isEnterInputsState(state: BaseInterpretedLaunchState): boolean {
     LaunchState.SUBMIT_SUCCEEDED,
   ].some(state.matches);
 }
+
+export function parseMappedTypeValue(value: InputValue): { key: string; value: string } {
+  try {
+    const mapObj = JSON.parse(value.toString());
+    const mapKey = Object.keys(mapObj)?.[0] || '';
```
Comment (anrusina) [inline packages/zapp/console/src/components/Launch/LaunchForm/utils.ts:198]: It's better to use `??` notation instead, as 
`const mapKey = Object.keys(mapObj)?.[0] ?? '';`

There are situations where || leads in such cases leads to very well hidden errors. I don't have an example from the top of my head. But I ca fins one if needed.

You are using it in few places here and in file above to default to an empty string value. Please replace.

**Review by anrusina — COMMENTED** [review]
haven't done smoke test. So doing only brief code review.
Good job overall,

- [ ] Can you please add screenshot or video with new Ui changes in it ([here is info how to do it](https://unionai.atlassian.net/wiki/spaces/FLYTE/pages/2097161/Pull+Request+PR))
- [ ] Please update PR description to include info what was done.  Remove `TL;DR` at the beginning, Remove Follow up issues - as it's needed only in case if your PR is not fully fixing the issue, and usualy uses a different ticket.
- [ ] Looks like Unit Tests are currently failing

File: packages/zapp/console/src/components/Launch/LaunchForm/MapInput.tsx (hunk 1 above)
Comment (james-union) [reply packages/zapp/console/src/components/Launch/LaunchForm/MapInput.tsx:76]: just followed the other parts of the code.. no specific reason

File: packages/zapp/console/src/components/Launch/LaunchForm/utils.ts (hunk 3 above)
Comment (james-union) [reply packages/zapp/console/src/components/Launch/LaunchForm/utils.ts:198]: fixed

File: packages/zapp/console/src/components/Launch/LaunchForm/inputHelpers/map.ts (hunk 2 above)
Comment (james-union) [reply packages/zapp/console/src/components/Launch/LaunchForm/inputHelpers/map.ts]: fixed

File: packages/zapp/console/src/components/Launch/LaunchForm/inputHelpers/map.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+function validate({ value, typeDefinition: { subtype } }: InputValidatorParams) {
+  if (!subtype) {
+    throw new Error(t('missingMapSubType'));
+  }
+  if (typeof value !== 'string') {
+    throw new Error('Value is not a string');
+  }
+  if (!value.toString().length) {
+    throw new Error('Value is required');
```
Comment (anrusina) [inline packages/zapp/console/src/components/Launch/LaunchForm/inputHelpers/map.ts:54]: Can you please add all strings mentioned in `Error(xxx)` in this file to string file.

Comment (james-union) [reply packages/zapp/console/src/components/Launch/LaunchForm/inputHelpers/map.ts]: fixed

**Review by anrusina — DISMISSED** [review]
Code and test drop looks Ok to me, with all things considered.
I'm fine to approve it from Techincal point of view. 
Just not sure if we agreed that it's aligns with Haytham and Ketan vision

**Review by anrusina — DISMISSED** [review]

**Review by anrusina — APPROVED** [review]

**Comment by james-union** [comment]
@jsonporter @anrusina Please review this again. I've added multiple keys ability for mapped type. The only issue is that I'm using `onBlur` on the keyInput and this rerenders the component that prevents onFocus for the valueInput first time.

**Comment by james-union** [comment]
Updated Video. It doesn't enable/disable the Launch Plan button but it submits the valid value (removing duplicates/empty key/values)

https://user-images.githubusercontent.com/105876962/176738220-fbc804ec-cac4-421e-8542-9c164922c8ec.mov

**Review by anrusina — APPROVED** [review]

**Comment by anrusina** [comment]
Failing test is not related to the PR. It is "fixed" separately in master: https://github.com/flyteorg/flyteconsole/pull/533

**Review by jsonporter — APPROVED** [review]
lgtm

**Pull request merged**